petab/v2/lint.py

```diff
@@ -0,0 +1,549 @@
+            measurement_df=problem.measurement_df,
+            mapping_df=problem.mapping_df,
+        )
+
+        actual = set(problem.parameter_df.index)
+        missing = required - actual
+        extraneous = actual - allowed
+
+        # missing parameters might be present under a different name based on
+        # the mapping table
+        if missing and problem.mapping_df is not None:
+            model_to_petab_mapping = {}
+            for map_from, map_to in zip(
+                problem.mapping_df.index.values,
+                problem.mapping_df[MODEL_ENTITY_ID],
+                strict=True,
+            ):
+                if map_to in model_to_petab_mapping:
+                    model_to_petab_mapping[map_to].append(map_from)
+                else:
+                    model_to_petab_mapping[map_to] = [map_from]
+            missing = {
+                missing_id
+                for missing_id in missing
+                if missing_id not in model_to_petab_mapping
+                or all(
+                    mapping_parameter not in actual
+                    for mapping_parameter in model_to_petab_mapping[missing_id]
+                )
+            }
+
+        if missing:
+            return ValidationError(
+                "Missing parameter(s) in the model or the "
+                "parameters table: " + str(missing)
+            )
+
+        if extraneous:
+            return ValidationError(
+                "Extraneous parameter(s) in parameter table: "
+                + str(extraneous)
+            )
+
+
+class CheckValidParameterInConditionOrParameterTable(ValidationTask):
+    """A task to check that all required and only allowed model parameters are
+    present in the condition or parameter table."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if (
+            problem.model is None
+            or problem.condition_df is None
+            or problem.parameter_df is None
+        ):
+            return
+
+        try:
+            assert_model_parameters_in_condition_or_parameter_table(
+                problem.model,
+                problem.condition_df,
+                problem.parameter_df,
+                problem.mapping_df,
+            )
+        except AssertionError as e:
+            return ValidationIssue(
+                level=ValidationIssueSeverity.ERROR, message=str(e)
+            )
+
+
+class CheckVisualizationTable(ValidationTask):
+    """A task to validate the visualization table of a PEtab problem."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if problem.visualization_df is None:
+            return
+
+        from petab.visualize.lint import validate_visualization_df
+
+        if validate_visualization_df(problem):
+            return ValidationIssue(
+                level=ValidationIssueSeverity.ERROR,
+                message="Visualization table is invalid.",
+            )
```

Comment:
For this single-error cases, there could already be some `SingleValidationTask` class like we discussed, e.g.

```suggestion
class CheckVisualizationTable(SingleValidationTask):
    """A task to validate the visualization table of a PEtab problem."""
    level = ValidationIssueSeverity.ERROR
    message = "Visualization table is invalid."

    def run(self, problem: Problem) -> ValidationIssue | None:
        if problem.visualization_df is None:
            return

        from petab.visualize.lint import validate_visualization_df

        return (not?) validate_visualization_df(problem)
        or
        return self.handle(validation_visualization_df(problem))
```

Reply:
I am not sure this will be more intuitive. People will always wonder whether `true` indicates passed or failed. Where the message is context-specific, this wouldn't really save anything. Will leave as is for now.

Reply:
Ah, I intended `self.handle` to return `ValidationIssue` or `None` like done elsewhere in this PR

Reply:
> Where the message is context-specific

Are you saying, an error message for `SingleValidationTasks` might be context-specific? Doesn't that contradict the idea of a `SingleValidationTask`?

Reply:
I meant that instead of `There are log-scaled parameters with negative bounds.`, we want to say `There are log-scaled parameters with negative bounds: parameter1, parameter42.`, so one would have to modify self.message again somewhere, and wouldn't really gain much.

Reply:
I see, but not if `self.handle` also accepts an optional level/message. Anyway, not important